m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+        metric.setTimestamp(System.currentTimeMillis() * Duration.NANOS_PER_MILLI);
+        metric.setMetricValue(metricValue);
+
+        return metric;
+    }
+
+    private void queueSizedMetric(SizedMetric sizedMetric) {
+        try {
+            metricQueue.put(sizedMetric);
+        } catch (InterruptedException e) {
+            LOG.warn(String.format("Interrupted queueing metric: %s", sizedMetric.getMetric().getName()));
+        }
+    }
+
+    private class Processor implements Runnable {
+        final Set<MetricTag> commonTags;
+        List<Metric> pendingMetrics = new ArrayList<>(freeBytes / 10);
+        int metricsSize = 0;
+
+        Processor(Set<MetricTag> commonTags) {
+            this.commonTags = commonTags;
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (!executor.isShutdown()) {
+                    // This `poll` call will block for at most the specified duration to take an item
+                    // off the queue. If we get an item, we append it to the queue to be flushed,
+                    // otherwise we flush what we have so far.
+                    // When this reporter is closed, shutdownNow will be called on the executor,
+                    // which will interrupt this thread and proceed to the `InterruptedException`
+                    // catch block.
+                    SizedMetric sizedMetric = metricQueue.poll(MAX_PROCESSOR_WAIT_UNTIL_FLUSH_MILLIS, TimeUnit.MILLISECONDS);
+
+                    // Drop metrics that came in after close
+                    if (sizedMetric == SizedMetric.CLOSE) {
+                        metricQueue.clear();
+                        break;
+                    }
+
+                    if (sizedMetric != null) {
+                        flushMetricIteration(sizedMetric);
+                        continue;
+                    }
+
+                    // If we don't get any new metrics after waiting the specified time,
+                    // flush what we have so far.
+                    if (!pendingMetrics.isEmpty()) {
+                        flushMetricIteration(SizedMetric.FLUSH);
+                    }
+                }
+            } catch (InterruptedException e) {
+                // Don't care if we get interrupted - the finally block will clean up
+            } finally {
+                flushRemainingMetrics();
+
+                // Always arrive at phaser to prevent deadlock
+                phaser.arrive();
+            }
+        }
+
+        private void flushMetricIteration(SizedMetric sizedMetric) {
+            Metric metric = sizedMetric.getMetric();
+
+            if (sizedMetric == SizedMetric.FLUSH) {
+                flush(pendingMetrics, commonTags);
+
+                return;
+            }
+
+            int size = sizedMetric.getSize();
+
+            if (metricsSize + size > freeBytes) {
+                flush(pendingMetrics, commonTags);
+                metricsSize = 0;
```

Comment:
you've already flushed the metric by the time you reach here, do you need a return after this line?